src/ConstantInterval.cpp

```diff
@@ -0,0 +1,643 @@
+    return a.max_defined && b.min_defined && a.max < b.min;
+}
+
+bool operator<=(const ConstantInterval &a, int64_t b) {
+    return a.max_defined && a.max <= b;
+}
+bool operator<(const ConstantInterval &a, int64_t b) {
+    return a.max_defined && a.max < b;
+}
+
+bool operator<=(int64_t a, const ConstantInterval &b) {
+    return b.min_defined && a <= b.min;
+}
+bool operator<(int64_t a, const ConstantInterval &b) {
+    return b.min_defined && a < b.min;
+}
+
+void ConstantInterval::cast_to(const Type &t) {
+    if (!t.can_represent(*this)) {
+        // We have potential overflow or underflow, return the entire bounds of
+        // the type.
+        ConstantInterval type_bounds;
+        if (t.is_int()) {
+            if (t.bits() <= 64) {
+                type_bounds.min_defined = type_bounds.max_defined = true;
+                type_bounds.min = ((int64_t)(-1)) << (t.bits() - 1);
+                type_bounds.max = ~type_bounds.min;
+            }
+        } else if (t.is_uint()) {
+            type_bounds.min_defined = true;
+            type_bounds.min = 0;
+            if (t.bits() < 64) {
+                type_bounds.max_defined = true;
+                type_bounds.max = (((int64_t)(1)) << t.bits()) - 1;
+            }
+        }
+        // If it's not int or uint, we're setting this to a default-constructed
+        // ConstantInterval, which is everything.
+        *this = type_bounds;
+    }
+}
+
+ConstantInterval ConstantInterval::operator-() const {
+    ConstantInterval result;
+    if (min_defined && min != INT64_MIN) {
+        result.max_defined = true;
+        result.max = -min;
+    }
+    if (max_defined) {
+        result.min_defined = true;
+        result.min = -max;
+    }
+    return result;
+}
+
+ConstantInterval ConstantInterval::bounds_of_type(Type t) {
+    return cast(t, ConstantInterval::everything());
+}
+
+ConstantInterval operator+(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.min_defined &&
+                         b.min_defined &&
+                         add_with_overflow(64, a.min, b.min, &result.min);
+
+    result.max_defined = a.max_defined &&
+                         b.max_defined &&
+                         add_with_overflow(64, a.max, b.max, &result.max);
+    return result;
+}
+
+ConstantInterval operator-(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.min_defined &&
+                         b.max_defined &&
+                         sub_with_overflow(64, a.min, b.max, &result.min);
+    result.max_defined = a.max_defined &&
+                         b.min_defined &&
+                         sub_with_overflow(64, a.max, b.min, &result.max);
+    return result;
+}
+
+ConstantInterval operator/(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+
+    result.min = INT64_MAX;
+    result.max = INT64_MIN;
```

Comment:
Shouldn't these be a.min and a.max? Otherwise I don't think this operator ever returns a bounded interval

Reply:
Oh, nevermind, I misread. Ignore.